app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -377,6 +426,135 @@ private int initializeAppResources(CommCareApp app) {
+            records.add(r);
+        }
+        Collections.sort(records, new Comparator<ApplicationRecord>() {
+
+            @Override
+            public int compare(ApplicationRecord lhs, ApplicationRecord rhs) {
+                return lhs.getDisplayName().compareTo(rhs.getDisplayName());
+            }
+
+        });
+        return records;
+    }
+
+    /**
+     * @return all ApplicationRecords that are installed and NOT archived
+     */
+    public ArrayList<ApplicationRecord> getVisibleAppRecords() {
+        ArrayList<ApplicationRecord> visible = new ArrayList<>();
+        for (ApplicationRecord r : getInstalledAppRecords()) {
+            if (!r.isArchived()) {
+                visible.add(r);
+            }
+        }
+        return visible;
+    }
+
+    /**
+     * @return  all ApplicationRecords that are installed AND are not archived AND have MM verified
+     */
+    public ArrayList<ApplicationRecord> getReadyAppRecords() {
+        ArrayList<ApplicationRecord> ready = new ArrayList<>();
+        for (ApplicationRecord r : getVisibleAppRecords()) {
+            if (r.resourcesValidated()) {
+                ready.add(r);
+            }
+        }
+        return ready;
+    }
+
+    public boolean visibleAppsPresent() {
+        return getVisibleAppRecords().size() > 0;
+    }
+
+    /**
+     * @return the list of all installed apps as an array
+     */
+    public ApplicationRecord[] appRecordArray() {
+        ArrayList<ApplicationRecord> appList = CommCareApplication._().getInstalledAppRecords();
+        ApplicationRecord[] appArray = new ApplicationRecord[appList.size()];
+        int index = 0;
+        for (ApplicationRecord r : appList) {
+            appArray[index++] = r;
+        }
+        return appArray;
+    }
+
+    /**
+     * @param index an index into the global table of installed apps
+     * @return the ApplicationRecord that corresponds to the given index in the global list of
+     * installed apps
+     */
+    public ApplicationRecord getAppAtIndex(int index) {
+        ArrayList<ApplicationRecord> currentApps = getInstalledAppRecords();
+        if (index < 0 || index >= currentApps.size()) {
+            return null;
+        } else {
+            return currentApps.get(index);
+        }
+    }
+
+    /**
+     * @param uniqueId - the uniqueId of the ApplicationRecord being sought
+     * @return the ApplicationRecord corresponding to the given id, IF and only if it is a "ready"
+     * app. Otherwise, return null
+     */
+    private ApplicationRecord getAppById(String uniqueId) {
+        for (ApplicationRecord r : getReadyAppRecords()) {
+            if (r.getUniqueId().equals(uniqueId)) {
+                return r;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Completes a full uninstall of the CC app that the given ApplicationRecord represents
+     */
+    public void uninstall(ApplicationRecord record) {
+        CommCareApp app = new CommCareApp(record);

```

Comment:
@ctsims Your comment on this got removed because I moved this method, so responding here- This is actually not a problem, at least not by all evidence thus far. There is currently no notion of unseating an app other than to seat a different app. Since CommCareApplication._().getCurrentApp() is the hook through which all parts of the app check what the currently-seated app is, calling initializeAppResources(CommCareApp), which overrides the currentApp field in CommCareApplication with the app passed in, is sufficient for both unseating the old and seating the new. Since the CommCareApp object is something that is recreated from the ApplicationRecord each time an app is seated, there isn't anything obvious to be done when "unseating" an app.

If you know of or suspect any weird behavior that this will cause, definitely tell me because if so, it will cause problems in many more places than just this. But from my understanding of how the CommCareApp/CommCareApplication code works this should be OK. And it's being done often and in many places, without hiccup thus far.